index.rst

```diff
@@ -15,15 +15,123 @@ Abstract
-
-.. Make in-text citations with: :cite:`bibkey`.
-.. Uncomment to use citations
-.. .. rubric:: References
-.. 
-.. .. bibliography:: local.bib lsstbib/books.bib lsstbib/lsst.bib lsstbib/lsst-dm.bib lsstbib/refs.bib lsstbib/refs_ads.bib
-..    :style: lsst_aa
+
+Some history
+============
+
+Alert Production Database (APDB) development was focused until recently on satisfying performance requirements of the AP Pipeline.
+Currently two implementation of APDB exist, one is based of SQL backend (PostgreSQL) and another based on NoSQL `Apache Cassandra`_.
+SQL backend exists solely for testing purposes, it was shown early on that a single-server RDBMS cannot satisfy AP performance requirements (`DMTN-113`_).
+Cassandra backend was shown to scale reasonably (`DMTN-156`_, `DMTN-184`_), currently we are in the process of setting up Cassandra cluster at USDF for further testing.
+
+APDB can only be used by AP pipeline, for daytime processing and serving user queries the data from APDB have to be copied to Prompt Products Database (PPDB).
+Similarly, the data produced during daytime processing in PPDB have to be copied to APDB for use during the next night.
+PPDB will be backed by RDBMS server (PostgreSQL) which allows arbitrary queries, though it is not clear yet that performance of a single-server instance is adequate for this purpose.
+The data in APDB and PPDB have to be eventually consistent, which makes these two separate systems one distributed database (at least in logical sense).
+
+The design of APDB/PPDB data transfers was the subject of an early `PPDB architecture meeting`_, though at that meeting we mostly considered replication of DIA tables (DIAObject, DIASource, and DIAForcedSource).
+More recent e-mail exchange provided additional details for how non-DIA tables (SSObject, SSSource, MPCORB) are populated and replicated.
+Here is a list of facts that summarizes AP pipeline interaction with APDB:
+
+- AP detects DIASources and generates DIAObjects from the 12-month history of DIASources and DIAForcedSources, to retrieve history it runs spatial/temporal queries on APDB which cover current detector+visit region.
+- If AP can associate DIASource with a known SSObject, it will read 12-month history of SSSources, update SSObject information and include SSObject and SSSources into the generated alert (and it will not make DIAObject).
+- To associate DIASource with a SSObject, it will need to read known SSObjects from APDB corresponding to the same detector+visit region.
+- All newly created and updated records for DIASource, SSSource, DIAObject or SSObject will be written to APDB.
+
+During daytime processing, which uses data from PPDB, some DIASources can be re-associated to either existing or new SSObjects, this can also cause removal of DIAObjects that have no matching DIASources.
```

Comment:
We wanted daytime processing to use the PPDB to ensure that it was working with filtered, releasable data.  Would it help avoid two-way propagation if we could have daytime processing use the APDB instead?

Reply:
(This might mean ensuring filtering at the input of the daytime processing, assuming that it is impossible to pre-filter all APDB entries.)

Reply:
Cassandra APDB schema is optimized for AP pipeline queries, I'm sure it will be super inefficient for daytime processing (and I'm not sure at this point if we can scale PPDB to provide reasonable performance). OTOH, I already have to duplicate recent APDB data for the purpose of replication (which is time-based), it may be possible to use that copy for daytime processing. My recent idea is that filtering and replication should go through a separate database (most likely relational), but I can try to think if filtering can be done on APDB too.